### PR TITLE
Adds a number of items

### PR DESCRIPTION
**Pull request by jlaura**
1.  CUDA coupled decomposition - this might still be having issues on very large images on the GPU, but works for our test cases.
1. Conditional CUDA dependency (inside __init__.py) and patched into the root namespace
  This PR incorporates CUDA support into autocnet. By default, the cuda support will not load. If the user calls autocnet.cuda(enable=True) the pyCudaSift library will attempt to import and a GPU will be allocated (the above func has a second kmart gpu that is the device id for multi-device systems. Once loaded, the GPU enabled funcs are patched into the Node and Edge classes where appropriate.
1. Refactor of stageful object: fundamental matrix and homography
1. Refactor attributes: small (in total size) attributes into the graph/edge/node attribute dict for improved serialization
1. Added __eq__ magic methods to edges/nodes to allow for edge == edge, node == node, and finally, CandidateGraph == CandidateGraph to support easier testing on
1. Added serialization/deserialization capability to save a project out to disk.  The project saves (overwrites) to an arbitrarily named zip file.  The output is pseudo opaque.  If the user wants to dig in, they can.  JSON inside is human readable, NPY arrays are not.
1. Removal of extraneous items.

### CODE REVIEW

**Comment by jlaura** [comment]
@Kelvinrr When you get a chance, can you checkout the two test failures that are occurring due to the MST computation?  I am not sure if I have made a change in our API that is causing the issue or if something else is going on.  Thanks!

**Comment by Kelvinrr** [comment]
Getting error: 

""""
  File "/Users/kelvinrodriguez/Repos/autocnet/autocnet/graph/network.py", line 10, in <module>
    from plio.io import io_hdf, io_json, io_autocnetgraph
ImportError: cannot import name 'io_autocnetgraph'
""""

`io_autocnetgraph` doesn't exist in Plio's github. Did you make the changes on your own copy and never published them?

**Comment by jlaura** [comment]
The error in Travis is due to the F-matrix returning using the 7 point algorithm (a 9x3 F matrix).  A new PR should be issued handling the ambiguity in that solution.  Move generally, should we even allow for F estimation when so few correspondences have been identified?

**Comment by jlaura** [comment]
The most recent addition  closes #178 and adds the ability to serialize to .npy format.  Also includes some improvements to test coverage.

**Comment by Kelvinrr** [comment]
Geez, finally got through this monstrous PR...

Automatic testing fails because of typical Travis CI shenanigans, running `pytest` locally shows all tests passing. 

Comments are inline, will merge as soon I see responses on those. Responses along the lines of: "that is a stupid question and you're stupid for asking" are totally fine.

**Comment by jlaura** [comment]
@Kelvinrr Am I missing inline comments?  Also, check the coverage numbers also including the functional tests.  If I omit those we lose ~10% coverage.  I see a drop 89% (still not good/mergable, but not so dire.)

```bash
---------- coverage: platform darwin, python 3.5.2-final-0 -----------
Name                                            Stmts   Miss  Cover   Missing
-----------------------------------------------------------------------------
autocnet/camera/__init__.py                         0      0   100%
autocnet/camera/camera.py                          39      3    92%   5-6, 163
autocnet/camera/utils.py                           12      0   100%
autocnet/cg/__init__.py                             0      0   100%
autocnet/cg/cg.py                                  74      0   100%
autocnet/control/__init__.py                        0      0   100%
autocnet/control/control.py                        66      0   100%
autocnet/examples/__init__.py                      12      0   100%
autocnet/graph/__init__.py                          0      0   100%
autocnet/graph/edge.py                            177     22    88%   61, 63-64, 82-83, 96, 112, 119, 127, 150-151, 197, 256-257, 319, 334-337, 425, 460, 484
autocnet/graph/markov_cluster.py                   41     38     7%   46-108
autocnet/graph/network.py                         199     10    95%   77, 80, 106, 130-131, 169, 299, 390, 430, 531
autocnet/graph/node.py                            192     28    85%   89-90, 93-94, 120, 137, 165, 294, 336-337, 342-346, 370-371, 449-466, 509
autocnet/io/__init__.py                             0      0   100%
autocnet/io/keypoints.py                           46      0   100%
autocnet/io/network.py                             62      2    97%   24, 66
autocnet/matcher/__init__.py                        0      0   100%
autocnet/matcher/add_depth.py                      21      1    95%   50
autocnet/matcher/ciratefi.py                      230     17    93%   102-104, 117, 256, 265, 269, 280, 318-322, 453, 462, 466, 474, 659
autocnet/matcher/cpu_decompose.py                 116     10    91%   74, 79, 98, 156, 158, 173, 195-197, 206
autocnet/matcher/feature.py                        44      1    98%   127
autocnet/matcher/feature_extractor.py              33      1    97%   72
autocnet/matcher/feature_matcher.py                22      2    91%   55, 60
autocnet/matcher/health.py                         16      6    62%   16, 31-35
autocnet/matcher/naive_template.py                 20      2    90%   48, 56
autocnet/matcher/outlier_detector.py              116      2    98%   89, 237
autocnet/matcher/subpixel.py                       25      3    88%   49, 80-81
autocnet/matcher/suppression_funcs.py              13      0   100%
autocnet/transformation/__init__.py                 0      0   100%
autocnet/transformation/decompose.py               65      2    97%   18, 31
autocnet/transformation/fundamental_matrix.py      64     31    52%   11-12, 45-54, 85-89, 113-114, 154-161, 172-173, 179, 199-207
autocnet/transformation/homography.py              40      9    78%   7-8, 34, 36, 96-101
autocnet/utils/__init__.py                          0      0   100%
autocnet/utils/observable.py                       34      6    82%   47, 68, 79-81, 85
autocnet/utils/utils.py                           105      8    92%   51, 122, 219-220, 337, 358-360
autocnet/vis/__init__.py                            0      0   100%
-----------------------------------------------------------------------------
TOTAL                                            1884    204    89%
```

**Comment by Kelvinrr** [comment]
Im an idiot, I had these comments since yesterday but I think you couldn't see them? I didn't realize I had to hit a submit button on the review comments. If thats true, then that means nobody has ever seen some of the comments on other peer reviews 😅 

Also, you are correct about the tests, the ones I ran didn't include the functional tests.

File: autocnet/__init__.py [hunk 1]
```diff
@@ -17,4 +9,48 @@ def get_data(filename):
...
+    dirname = os.path.join(os.path.dirname(packagdir), 'data')
+    fullname = os.path.join(dirname, filename)
+    return fullname
+
+def cuda(enable=False, gpu=0):
+    # Classes/Methods that can vary if GPU is available
+    from autocnet.graph.node import Node
+    from autocnet.graph.edge import Edge
+    if enable:
+        print('Enabling CUDA')
```
Comment (jlaura) [inline autocnet/__init__.py]: I wasn't sure on this.  Do you think setting a module level (really project level) variable is enough?  Then warn if enabling fails.

File: autocnet/graph/network.py [hunk 2]
```diff
@@ -7,24 +7,23 @@
...
 
-from plio.io import io_hdf
-from plio.io import io_json
+from plio.io import io_hdf, io_json
 from plio.utils import utils as io_utils
 from plio.io.io_gdal import GeoDataset
 from autocnet.graph import markov_cluster
 from autocnet.graph.edge import Edge
 from autocnet.graph.node import Node
+from autocnet.io import network as io_network
```
Comment (jlaura) [inline autocnet/graph/network.py]: Intentional.  I am torn here.  Things like plio.io_gdal.GeodataSet are really generic.  Things like this io_network are super specific to this project.  I am struggling with where to break the coupling.  Any suggestions?

File: autocnet/io/keypoints.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    descriptors : ndarray
+                  of feature descriptors
+
+    out_path : str
+               PATH and filename to save the features
+    """
+    np.savez(out_path, descriptors=descriptors,
+             keypoints=keypoints,
+             keypoints_idx=keypoints.index,
+             keypoints_columns=keypoints.columns)
```
Comment (jlaura) [inline autocnet/io/keypoints.py]: See the above, but basically, I am soliciting input on this.  Where does the I/O break occur?  Is plio an acceptable dependency where all I/O, no matter how module specific, should be in there?  That creates a hard plio dependency, which I do not know if I like.

File: autocnet/matcher/cpu_decompose.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+        fl.train()
+        matches = fl.query(bd, b['node_id'], k, index=bidx)
+        if self.matches is None:
+            self.matches = matches
+        else:
+            df = self.matches
+            self.matches = df.append(matches,
+                                     ignore_index=True,
+                                     verify_integrity=True)
+        fl.clear()
```
Comment (jlaura) [inline autocnet/matcher/cpu_decompose.py]: 👍 I'll get this pulled out.

**Comment by jlaura** [comment]
I worked some on increasing test coverage.  I would prefer to continue to improve coverage (tonight or tomorrow probably) before merging this.  Your point re: testing is 100% valid.

Also, definitely no dumb review questions.  I know that I need to break down the PR more next time, even if just 'piece-mealing' the code in when I get a second, so apologies there.

**Comment by jlaura** [comment]
This block yanks unused modules, refactors out all observables, adds new metric for the F matrix, updates IO for missing attributes, removes the redundant mono-matches, and gets the test coverage back up.  Phewww! 

Let me know about additional questions, etc.  I would love to get this closed ASAP so that we can all get back to small PRs.

File: autocnet/graph/network.py (hunk 2 above)
Comment (Kelvinrr) [inline autocnet/graph/network.py]: Yeah, I see what you mean. I imagine as long as this is the only place `io_network` and `io_autocnetgraph` are being used is here and has no use case elsewhere, then keeping it part of `autocnet` seems to be the way to go. If that is the choice you make, worth deleting them from `plio`

File: .travis.yml [hunk 5]
```diff
@@ -54,7 +54,6 @@ install:
   # Development installation
   - conda install pytest pytest-cov coverage sh anaconda-client
   - pip install coveralls
-  - python runipynbs.py
```
Comment (Kelvinrr) [inline .travis.yml]: Not interested in testing notebooks anymore?

**Comment by jlaura** [comment]
Nope. Run ipynb is deprecated and a pin to maintain. Would rather use a
different repo for examples.

**Comment by Kelvinrr** [comment]
Yay! I think this is good. Maybe in the future we should avoid these huge PR's. They are a pain to review and can be easy to accidentally miss something in the great pile of changes.

**Comment by jlaura** [comment]
Agree 100% this big is bad.

**Pull request merged**